test/cmd/packet-verifier/packet-verifier.go

```diff
@@ -0,0 +1,272 @@
+			parentLink, err := netlink.LinkByIndex(link.Attrs().ParentIndex)
+			if err != nil {
+				fmt.Printf("unable to get parent link %d. Error %+v", link.Attrs().ParentIndex, err)
+				os.Exit(1)
+			}
+			trunkDevToMonitor := eniConfig{name: parentLink.Attrs().Name, shouldVerifyVlanTag: true}
+			enis = append(enis, trunkDevToMonitor)
+		} else {
+			// find the eni to monitor associated with the pod
+			rules, err := netlink.RuleList(netlink.FAMILY_V4)
+			if err != nil {
+				fmt.Printf("unable to get ip rules due to %+v", err)
+				os.Exit(1)
+			}
+			for _, rule := range rules {
+				// Find the ENI in the route table associated with the pod
+				if rule.Src != nil && net.ParseIP(ipToMonitor).Equal(rule.Src.IP) {
+					routeFilter := &netlink.Route{
+						Table: rule.Table,
+					}
+					routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, routeFilter, netlink.RT_FILTER_TABLE)
+					if err != nil {
+						fmt.Printf("unable to get routes based on iprules %+v", err)
+						os.Exit(1)
+					}
+
+					for _, route := range routes {
+						parentLink, err := netlink.LinkByIndex(route.LinkIndex)
+						if err != nil {
+							fmt.Printf("unable to get parent ENI for the link %+v", err)
+							os.Exit(1)
+						}
+						eniToMonitor := eniConfig{name: parentLink.Attrs().Name}
+						enis = append(enis, eniToMonitor)
+						break
+					}
+				}
+			}
+			// if explicit route is not round then pod has to be associated with $device
+			if len(enis) == 0 {
+				eniToMonitor := eniConfig{name: device}
+				enis = append(enis, eniToMonitor)
+			}
+		}
+	}
+
+	fmt.Printf("ENIs to monitor: %+v\n", enis)
+
+	err = monitorPacketOnInterfaces(ipToMonitor, vlanIDToMonitor, enis)
+	if err != nil {
+		fmt.Printf("unable to verify packets on the interface %+v ", err)
+		os.Exit(1)
+	}
+
+	fmt.Println("Successfully verified all the interfaces.")
+}
+
+// monitorPacketOnInterfaces invokes monitorPackets for each interface
+func monitorPacketOnInterfaces(ipAddress string, vlanIDToMonitor int, enis []eniConfig) error {
+
+	for _, iface := range enis {
+		fmt.Printf("Verifying interface: %+v\n", iface)
+		err := monitorPackets(ipAddress, vlanIDToMonitor, iface)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// monitorPackets monitors the packets on the interfaces
+func monitorPackets(ipAddress string, vlanIDToMonitor int, iface eniConfig) error {
+	handle, err := pcap.OpenLive(iface.name, snapshot_len, promiscuous, timeout)
+	if err != nil {
+		return err
+	}
+	defer handle.Close()
+	var srcPacketsProcessed int
+	var dstPacketsProcessed int
+
+	// Use the handle as a packet source to process all packets
+	packetSource := gopacket.NewPacketSource(handle, handle.LinkType())
+	for packet := range packetSource.Packets() {
+		fmt.Printf("packet: %v\n", packet)
+
+		network := packet.Layer(layers.LayerTypeIPv4)
+		if network != nil {
+			if iface.shouldCheckSrc && packet.NetworkLayer().NetworkFlow().Src().String() != ipAddress &&
```

Comment:
parse `ipAddress` earlier and use `net.IP.Equal()` here.  It will be (much) faster, and avoid non-canonical strings causing spurious differences.

(Same throughout this function)

Reply:
Will update this, thanks for the suggestion.